keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,289 @@
+   API servers is one.
+
+### Garbage collection of the StorageVersion objects of removed resources
+
+During bootstrap, the kube-apiserver
+* gets the list of all participating API servers,
+* gets all the StorageVersion objects,
+* removes the StorageVersion object if none of its CandidateVersions.APIServerID
+  is a participating API server.
+
+### Using the selected storage version when serializing data for etcd
+
+Depending on how strict the guarantee we want to provide, we have two
+alternative designs.
+
+1. strict guarantee: kube-apiserver always use the selected storage version
+
+To guarantee this, kube-apiserver needs to add a transaction condition to the
+write requests, to make sure the version it uses to serialize the API object is
+the selected storage version recorded in the API server.
+
+The performance cost is small, because all write operations from the
+kube-apiserver to etcd are transactional already.  For example, kube-apiserver
+checks if key does not already exist before committing a create operation. As
+another example, kube-apiserver checks that the resourceVersion is the latest
+before committing an update operation. Adding another transaction condition does
+not add extra RPC call.
+
+If the transaction fails, the kube-apiserver gets the latest selected storage
+version from etcd, re-serializes the data and retries the commit. The
+kube-apiserver will use this latest selected storage version in the following
+operations.
+
+2. loose guarantee: kube-apiserver has a very high probability to use the
+   selected storage version 1 minute after the version is selected in etcd.
+
+To guarantee this, the kube-apiserver can rely on a watch channel to deliver
+the latest storage version within 1 minute. 
+
+This saves the performance cost for the server, but shifts the complexity to the
+clients. For example, the storage migrator should start migration 1 minute after
+detecting the storage version change.
+
+
+### Updating the discovery document
+
+Currently, the apiservers expose the storage version via the
+[storageVersionHash][] field in the discovery document. The storage versions for
+built-in resources have been static. With this KEP, the kube-apiserver needs to
+dynamics update the storage version hash in the discovery document when the it
+gets the latest selected storage version from etcd.
+
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
+
+### CRDs
+
+Today, a [apiextension-apiserver][] uses informer to watch for the latest
+change to the CRD spec, and [builds][] new storage encoder with the latest storage
+version.
+
+The CRDs do not need the `StorageVersion` API to vote for a selected storage
+version. The storage version in the CRD spec is the "selected storage version".
+
+The current implementation of the apiextension-apiserver already provides the
+"loose guarantee".
```

Comment:
Do we have any signals from users that it's not enough?
If not, maybe it's not worth changing that (and maybe that also means it should be good enough for builtin resources).

Reply:
No, I haven't heard any complain.

I'm willing to pursue the "loose guarantee" path if all reviewers think it's good enough.